source/common/tls/default_tls_certificate_selector.cc

```diff
@@ -180,24 +180,20 @@ DefaultTlsCertificateSelector::findTlsContext(
   const Ssl::TlsContext* candidate_ctx = nullptr;
   Ssl::OcspStapleAction ocsp_staple_action;
 
+  int cec_count = client_ecdsa_capabilities.size();
+
   auto selected = [&](const Ssl::TlsContext& ctx) -> bool {
     auto action = ocspStapleAction(ctx, client_ocsp_capable);
     if (action == Ssl::OcspStapleAction::Fail) {
       // The selected ctx must adhere to OCSP policy
       return false;
     }
-    // if the client is ECDSA-capable and the context is ECDSA, we check if it is capable of
-    // handling the curves in the cert in a given TlsContext
-    if (client_ecdsa_capabilities.has_value() && ctx.is_ecdsa_) {
-      bssl::UniquePtr<EVP_PKEY> public_key(X509_get_pubkey(ctx.cert_chain_.get()));
-      // we're doing this with a guard that the `ctx` is ECDSA - this should be safe
-      const EC_KEY* ecdsa_public_key = EVP_PKEY_get0_EC_KEY(public_key.get());
-      ASSERT(ecdsa_public_key != nullptr);
-      const EC_GROUP* ecdsa_group = EC_KEY_get0_group(ecdsa_public_key);
-      const int ecdsa_curve_nid = EC_GROUP_get_curve_name(ecdsa_group);
-      // if we have a matching curve NID in our client capabilities, return `true`
-      const std::vector<int> cec_vec = client_ecdsa_capabilities.value();
-      if (std::find(cec_vec.begin(), cec_vec.end(), ecdsa_curve_nid) != cec_vec.end()) {
+    // If the client is ECDSA-capable and the context is ECDSA, we check if it is capable of
+    // handling the curves in the cert in a given TlsContext.
+    if (cec_count != 0 && ctx.ec_group_curve_name_ != 0) {
```

Comment:
I don't think either of these checks are needed. If `cec_count == 0` then nothing will be found in the empty vector, so the check isn't needed. If `ec_group_curve_name_ == 0`, it won't be found in the vector because that's not a curve NID.